verify.js

```diff
@@ -86,92 +51,134 @@ module.exports = function (jwtString, secretOrPublicKey, options, callback) {
 
   var header = decodedToken.header;
 
-  if (!~options.algorithms.indexOf(header.alg)) {
-    return done(new JsonWebTokenError('invalid algorithm'));
+  if(header.alg ==='none' && decodedToken.signature) {
+    return done(new JsonWebTokenError('invalid token: unsigned but signature is present'));
   }
 
-  var valid;
+  var getSecret;
 
-  try {
-    valid = jws.verify(jwtString, header.alg, secretOrPublicKey);
-  } catch (e) {
-    return done(e);
+  if(typeof secretOrPublicKey === 'function') {
+    if(!callback) {
+      return done(new JsonWebTokenError('verify must be called asynchronous if secret or public key is provided as a callback'));
+    }
+
+    getSecret = secretOrPublicKey;
+  }
+  else {
+    getSecret = function(header, callback) {
+      return callback(undefined, secretOrPublicKey);
+    };
   }
 
-  if (!valid)
-    return done(new JsonWebTokenError('invalid signature'));
+  return getSecret(header, function(err, secretOrPublicKey) {
```

Comment:
I'd add the claims as well, they can be used by the function to use the issuer for JWKs: https://auth0.com/docs/jwks

Reply:
I'd advise against that. The payload shouldn't be used before it token has been verified. Using the issuer is especially dangerous. The risk is that someone will use it to fetch the key from that issuer. A hacker can then easily make a token that will always pass. I don't see a use-case for it either as the client should already know which issuer to use for instance with a config setting.

Reply:
Let's say I have an API, that API accepts tokens from 2 issuers, legacy authorization server and new authorization system, both of them hase a policy of certificate rotation, using JWKs to offer their public keys. The API, as a token verifier, needs to check their public keys in order to verify the signature, how would the API know that is the public key "now for this incoming JWT for this issuer"?

The API could use the `getSecret` function, asserting that the domain (+path?+protocol?...) for the issuer is in their list (like verifying the issuer) to prevent SSRF attacks, after the API could call the JWK endpoint to get the public key (probably with a cache), then it will filter by `kid` to select the right one.

> The risk is that someone will use it to fetch the key from that issuer. A hacker can then easily make a token that will always pass
What you fetch from the issuer is the public key. What is used to sign is the private key, that is not offered by the issuer.

Reply:
We are talking security here and imo that should be the priority. The payload has not been verified yet and is therefor not trustworthy. Using it anyway can open a host of security risks.

The use-case you describe can be solved in several other ways that won't compromise security:
* the legacy system can use a fixed key, and it can be easily identified by the `kid`
* if key rotation is required, one could try to fetch from both auth servers, only one will give a key. It can be cached so no performance is lost
* one could use a different URL or endpoint that can be used to select the server

Imo the payload should be made only available when decode returns, and only if it has been verified. Using the payload before that for any purpose just isn't hygienic.

But in the end I'm not the boss of this package so it's just my opinion of course.

Reply:
Using the header or the payload has the same effect: you are using fields without being verified. In both use cases (load key by `kid` and/or by `issuer`) you are relaying on the not-verified information and you will check if that makes sense (it's verified) afterwards.

If someone modifies the header to set a different `kid`, what would happen? The consumer code on `getSecret` won't find the key, and the verification will fail.
In case of `issuer` it's the same, consumer code on `getSecret` will try to get the key, if it can't get it the verification will fail.

On `getSecret` code the consumer must use only the claims/headers needed to figure out the key to use to verify the token.

Reply:
Just adding my .02:  When using ID Tokens from the OpenID Connect, you have the `iss` in payload. Fetching the keys from that exact issuer (via its `well-known/openid-configuration`) is prudent, because the thing you are validating is that the id_token is signed by that very same issuer, and nothing else.  Since the issuer is part of the subjects identity, it does not open a security issue per se.  The authentication is about `sub@iss` always, and we should allow for that use case.